examples/server/public/index.html

```diff
@@ -0,0 +1,300 @@
+      padding: 1em;
+    }
+
+    main .content {
+      flex-grow: 1;
+    }
+
+    main .content > * {
+      overflow-y: auto;
+    }
+
+    footer {
+      font-size: smaller;
+      color: #999;
+    }
+
+    body {
+      max-width: 600px;
+      min-width: 300px;
+      line-height: 1.2;
+      margin: 0 auto;
+    }
+
+    p {
+      overflow-wrap: break-word;
+      word-wrap: break-word;
+      hyphens: auto;
+      margin-top: 0.5em;
+      margin-bottom: 0.5em;
+    }
+
+    form {
+      margin: 1em 0 0 0;
+      display: flex;
+      gap: 0.5em;
+      flex-direction: row;
+      align-items: center;
+    }
+
+    form > * {
+      padding: 4px;
+    }
+
+    form input {
+      flex-grow: 1;
+    }
+
+    form textarea {
+      flex-grow: 1;
+      min-height: 5em;
+      width: 100%;
+    }
+
+    fieldset {
+      width: 100%;
+      padding: 1em;
+    }
+
+    fieldset label {
+      margin: 0.5em 0;
+      display: block;
+    }
+  </style>
+
+  <script type="module">
+    import {
+      html, h, signal, effect, computed, render, useSignal, useEffect, useRef
+    } from '/index.js';
+
+    import { llamaComplete } from '/completion.js';
+
+    const transcript = signal([])
+    const chatStarted = computed(() => transcript.value.length > 0)
+
+    const chatTemplate = signal("{{prompt}}\n\n{{history}}\n{{bot}}:")
+    const settings = signal({
+      prompt: "This is a conversation between user and llama, a friendly chatbot.",
+      bot: "llama",
+      user: "User"
+    })
+
+    const temperature = signal(0.2)
+    const nPredict = signal(80)
+    const controller = signal(null)
+    const generating = computed(() => controller.value == null )
+
+    // simple template replace
+    const template = (str, map) => {
+      let params = settings.value;
+      if (map) {
+        params = { ...params, ...map };
+      }
+      return String(str).replaceAll(/\{\{(.*?)\}\}/g, (_, key) => template(params[key]));
+    }
+
+    // send message to server
+    const chat = async (msg) => {
+      if (controller.value) {
+        console.log('already running...');
+        return;
+      }
+      controller.value = new AbortController();
+      transcript.value = [...transcript.value, ['{{user}}', msg]];
+
+      const payload = template(chatTemplate.value, {
+        message: msg,
+        history: transcript.value.flatMap(([name, msg]) => `${name}: ${msg}`).join("\n"),
+      });
+
+      let currentMessage = '';
+      let history = transcript.value;
+
+      const params = {
+        prompt: payload,
+        n_predict: parseInt(nPredict.value),
+        temperature: parseFloat(temperature.value),
+        stop: ["</s>", template("{{bot}}:"), template("{{user}}:")],
+      }
+
+      await llamaComplete(params, controller.value, (message) => {
+        const data = message.data;
+        currentMessage += data.content;
+
+        transcript.value = [...history,["{{bot}}", currentMessage]];
+
+        if (data.stop) {
```

Comment:
I think this solves most of the whitespace issues and doesn't cause the server to reevaluate too much:

```suggestion
        if (data.stop) {
          transcript.value = [...history,["{{bot}}", currentMessage.trim()]];
```